coderd/workspaceupdates_test.go

```diff
@@ -0,0 +1,345 @@
+package coderd_test
+
+import (
+	"context"
+	"encoding/json"
+	"slices"
+	"strings"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/coderd"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/coderd/rbac/policy"
+	"github.com/coder/coder/v2/coderd/wspubsub"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestWorkspaceUpdates(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	ws1ID := uuid.New()
+	ws1IDSlice := tailnet.UUIDToByteSlice(ws1ID)
+	agent1ID := uuid.New()
+	agent1IDSlice := tailnet.UUIDToByteSlice(agent1ID)
+	ws2ID := uuid.New()
+	ws2IDSlice := tailnet.UUIDToByteSlice(ws2ID)
+	ws3ID := uuid.New()
+	ws3IDSlice := tailnet.UUIDToByteSlice(ws3ID)
+	agent2ID := uuid.New()
+	agent2IDSlice := tailnet.UUIDToByteSlice(agent2ID)
+	ws4ID := uuid.New()
+	ws4IDSlice := tailnet.UUIDToByteSlice(ws4ID)
+
+	ownerID := uuid.New()
+	memberRole, err := rbac.RoleByName(rbac.RoleMember())
+	require.NoError(t, err)
+	ownerSubject := rbac.Subject{
+		FriendlyName: "member",
+		ID:           ownerID.String(),
+		Roles:        rbac.Roles{memberRole},
+		Scope:        rbac.ScopeAll,
+	}
+
+	t.Run("Basic", func(t *testing.T) {
+		t.Parallel()
+
+		db := &mockWorkspaceStore{
+			orderedRows: []database.GetWorkspacesAndAgentsByOwnerIDRow{
+				// Gains a new agent
+				{
+					ID:         ws1ID,
+					Name:       "ws1",
+					JobStatus:  database.ProvisionerJobStatusRunning,
+					Transition: database.WorkspaceTransitionStart,
+					Agents: []database.AgentIDNamePair{
+						{
+							ID:   agent1ID,
+							Name: "agent1",
+						},
+					},
+				},
+				// Changes status
+				{
+					ID:         ws2ID,
+					Name:       "ws2",
+					JobStatus:  database.ProvisionerJobStatusRunning,
+					Transition: database.WorkspaceTransitionStart,
+				},
+				// Is deleted
+				{
+					ID:         ws3ID,
+					Name:       "ws3",
+					JobStatus:  database.ProvisionerJobStatusSucceeded,
+					Transition: database.WorkspaceTransitionStop,
+				},
+			},
+		}
+
+		ps := &mockPubsub{
+			cbs: map[string]pubsub.ListenerWithErr{},
+		}
+
+		updateProvider, err := coderd.NewUpdatesProvider(slogtest.Make(t, nil), ps, db, &mockAuthorizer{})
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			_ = updateProvider.Close()
+		})
+
+		sub, err := updateProvider.Subscribe(dbauthz.As(ctx, ownerSubject), ownerID)
+		require.NoError(t, err)
+		ch := sub.Updates()
+
+		update, ok := <-ch
```

Comment:
use `update := testutil.RequireRecvCtx(ctx, t, ch)` for this, with a context that times out after `testutil.WaitShort` --- this helps prevent bugs in the update provider from hanging the entire test suite.